test/server/guarddog_impl_test.cc

```diff
@@ -0,0 +1,156 @@
+#include "common/common/utility.h"
+
+#include "server/guarddog_impl.h"
+
+#include "test/mocks/common.h"
+#include "test/mocks/server/mocks.h"
+#include "test/mocks/stats/mocks.h"
+
+using testing::InSequence;
+using testing::NiceMock;
+
+namespace Server {
+
+/**
+ * Death test caveat: Because of the way we die gcov doesn't receive coverage
+ * information from the forked process that is checked for succesful death.
+ * This means that the lines dealing with the calls to PANIC are not seen as
+ * green in the coverage report. However, rest assured from the results of the
+ * test: these lines are in fact covered.
+ */
+class GuardDogDeathTest : public testing::Test {
+protected:
+  GuardDogDeathTest()
+      : config_kill_(1000, 1000, 100, 1000), config_multikill_(1000, 1000, 1000, 500),
+        time_point_(std::chrono::system_clock::now()) {}
+
+  NiceMock<Configuration::MockMain> config_kill_;
+  NiceMock<Configuration::MockMain> config_multikill_;
+  NiceMock<Stats::MockStore> fakestats_;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+class GuardDogMissTest : public testing::Test {
+protected:
+  GuardDogMissTest() : config_miss(1, 1000, 0, 0), config_mega(1000, 1, 0, 0) {}
+
+  NiceMock<Configuration::MockMain> config_miss;
+  NiceMock<Configuration::MockMain> config_mega;
+  NiceMock<Stats::MockStore> stats_store;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+TEST_F(GuardDogDeathTest, KillDeathTest) {
+  // Is it German for "The Function"? Almost...
+  auto die_function = [&]() -> void {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_kill_, time_source_);
+    auto unpet_dog = gd.createWatchDog(0);
+    gd.forceCheckForTest();
+    time_point_ += std::chrono::milliseconds(500);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.forceCheckForTest();
+  };
+  // Why do it this way? Any threads must be started inside the death test
+  // statement and this is the easiest way to accomplish that.
+  EXPECT_DEATH(die_function(), "");
+}
+
+TEST_F(GuardDogDeathTest, MultiKillDeathTest) {
+  auto die_function = [&]() -> void {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+    auto unpet_dog1 = gd.createWatchDog(0);
+    gd.forceCheckForTest();
+    auto unpet_dog2 = gd.createWatchDog(1);
+    gd.forceCheckForTest();
+    time_point_ += std::chrono::milliseconds(501);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.forceCheckForTest();
+  };
+  EXPECT_DEATH(die_function(), "");
```

Comment:
No, this is the quirk of the death test when there are threads in play: The thread must be started within the statement in EXPECT_DEATH.